### ISSUE TITLE
Multi-Quantile / Distributional Regression Support

### ISSUE THREAD

**Issue by jonaslandsgesell** [google-research/tabfm#31]
**Motivation**
Currently, TabFMRegressor focuses on point-estimate regression (minimizing MSE/RMSE). However, point estimates discard predictive uncertainty, which is critical for real-world tabular data plagued by heavy tails or heteroskedasticity (see Image 1 for an example of suboptimal modeling with point estimates).

We recently introduced ScoringBench ([scoringbench.com](https://scoringbench.com/)), a framework designed to evaluate tabular models under proper scoring rules like CRPS and interval scores. The theoretical motivation and empirical shifts in model rankings are detailed in our accompanying preprints:

[ScoringBench: A Benchmark for Evaluating Tabular Foundation Models with Proper Scoring Rules (arXiv:2603.29928)](https://arxiv.org/abs/2603.29928)

[Distributional Regression with Tabular Foundation Models: Evaluating Probabilistic Predictions via Proper Scoring Rules (arXiv:2603.08206)](https://arxiv.org/abs/2603.08206)

To rigorously evaluate TabFM's zero-shot uncertainty probabilistic predictive performance on this benchmark, the model needs the ability to output a predictive distribution rather than a single point estimate

**Proposed Feature: Multi-Quantile Regression**
We propose extending TabFM to support non-parametric distributional regression e.g. via multi-quantile estimation (as in tabiclv2):

API Extension: Allow users to request an array of target quantiles, e.g., predict_quantiles(X_test, quantiles=[0.01, ... 0.05, 0.25, ... 0.5, 0.75, 0.95]) (in a fine grid).

Loss Optimization: Adapt the output head or decoding strategy to optimize pinball loss across the specified spectrum simultaneously.

Since the integral of the pinball loss approximates the Continuous Ranked Probability Score (CRPS), this directly optimizes the model for proper scoring rules.

**Evaluation**
Adding this feature would allow us to benchmark TabFM on the ScoringBench suite, providing a direct, statistically rigorous comparison of TabFM's probabilistic performance against baselines like TabPFN and quantile XGBoost and many other conditional density estimators.

**Comment by JuanVargas** [comment]
"... We propose extending TabFM to support non-parametric distributional regression e.g. via multi-quantile estimation (as in tabiclv2)..."

Good idea. I also would like to see that added to TabFM.

**Comment by Whatsonyourmind** [comment]
Great direction — and since ScoringBench is the target, it's worth getting the CRPS↔pinball mapping *exact* rather than approximate. Three things that bite once the grid becomes a CRPS number you compare across models:

**1. The factor of 2 (and grid normalization).** The relation is exact, not just approximate: CRPS(F, y) = 2 ∫₀¹ QL_τ(F⁻¹(τ), y) dτ, so the mean pinball loss over a uniform τ-grid is a Riemann estimate of CRPS/2, not CRPS. When you rank TabFM against sample-CRPS baselines (TabPFN, quantile XGB) on ScoringBench, a dropped factor of 2 or an inconsistent grid normalization surfaces as a spurious level shift in the comparison. Cheapest to pin the discretized estimator to the exact identity up front.

**2. Independent pinball heads cross.** Optimizing q(τ_k) independently across a fine grid doesn't constrain monotonicity, so q(0.25) > q(0.75) happens — and a non-monotone quantile function makes the implied CDF (and the CRPS integral over it) ill-defined. Two standard fixes: a monotone output head, or post-hoc rearrangement (Chernozhukov–Fernández-Val–Galichon 2010), which is free and provably does not increase estimation error. The finer the grid, the more this matters.

**3. A uniform grid under-resolves the tails — which is the stated motivation.** The Riemann error concentrates where F⁻¹ is steep, i.e. the tails, and CRPS on heavy-tailed targets is dominated by exactly those. A uniform τ-grid systematically under-integrates the tail contribution, so the heteroskedastic / heavy-tail cases cited as the reason for this feature are the ones a coarse uniform grid scores least faithfully. A tail-denser grid (or trapezoidal weights on a uniform one) recovers most of it at near-zero cost.

Happy to sketch the rearrangement + exact-CRPS-from-grid piece against a `predict_quantiles` signature if it's useful.

**Comment by jonaslandsgesell** [comment]
Great Points!

**Training vs. Evaluation Objectives (The Factor of 2)**
A positive affine transformation of a strictly proper scoring rule is still a proper scoring rule with the same orientation (i.e., maximization/minimization). A factor of 2 may change the gradients and the interplay with the learning rate, but the overall minimizer is the same. Therefore, the training loss of the model is a modeling decision (where multipliers play the role of changing your effective learning rate).
With regards to evaluation, we evaluate all models on CRPS—strictly speaking, the energy beta = 1 version of it, and assuming constant densities in the bins induced e.g. by the quantiles. (See the [ScoringBench metrics implementation](https://github.com/jonaslandsgesell/ScoringBench/blob/master/scoringbench/metrics.py#L145)).

**Quantile Crossing & Post-Hoc Corrections**
Quantile crossing is a known problem. (You can read more on this in the TabICLv2 paper, [Appendix page 40 on "Quantile crossing correction"](https://arxiv.org/pdf/2602.11139); even for TabICLv2, the ScoringBench wrapper applies the sorting correction, though you could implement a more sophisticated method in the wrapper if you have other best practices. [Izbicki et al.](https://arxiv.org/abs/2603.26611) also describe that sorting is needed in their appendix).
While a monotone output head as an architectural decision makes sense, post-hoc fixes can be implemented in the wrapper if needed since the ScoringBench wrapper is flexible. Any post-hoc method coupled to a model can simply be viewed as a new effective model added to ScoringBench.

**Grid Resolution, Support Points & the Sampling Trade-off**
With regards to the ScoringBench wrapper, we are happy to exchange ideas about the placement of the support points of the PMF grid (perhaps in terms of quantiles). The implementation of the CRPS used in ScoringBench can be found [here](https://github.com/jonaslandsgesell/ScoringBench/blob/master/scoringbench/metrics.py#L145) (which uses the beta = 1.0 energy score equivalence).

Currently, the number of bins and their (quantile) placement are specified by the corresponding model wrappers (as a hyperparameter of their modeling decisions).
The ScoringBench wrapper allows you to choose whatever quantile discretization you want. Philosophically, currently we view the choice of discretization of the probability output as one hyperparameter of your model and we test that the implemented scoring rules do not change extremely when doubling the number of bins. 
You can choose to set quantiles denser in the tails if you wish, but the estimation error of quantiles increases in the tails—this is already true for [unconditional quantile estimators](https://en.wikipedia.org/wiki/Quantile#The_asymptotic_distribution_of_the_sample_median) without any machine learning involved. Consequently, you may face statistical sampling artifacts at extreme quantiles even though the numerical integration side "improves".

Dealing with "extreme tails" is a known limitation for scoring rules (Sebastian Lerch, Thordis L Thorarinsdottir, Francesco Ravazzolo, and Tilmann Gneiting. Forecaster’s dilemma:
extreme events and forecast evaluation. Statistical Science, pages 106–127, 2017. and also Jonas R. Brehmer and Kirstin Strokorb. Why scoring functions cannot assess tail properties. Electronic Journal of
Statistics, 13(2):4015– 4034, 2019.). Putting more focus on the tails may partially be addressed via using quantile-weighted versions of scoring rules (e.g., weighted CRPS). 

I am happy to also align or assist with you via private communication if you wish.

**Comment by Whatsonyourmind** [comment]
Thanks — that clears up all three, and the affine-invariance point is right: a·S + b with a > 0 preserves both propriety and the minimizer, so the factor of 2 is a training-loss / effective-LR choice, not a correctness issue. And since ScoringBench computes CRPS directly as the β = 1 energy score (constant-density bins), the eval side is already normalized — my caveat only bites for a *model wrapper* that self-reports "CRPS" from a raw mean-pinball grid without that factor, which your pipeline sidesteps by scoring centrally. Good.

On the tails, your Brehmer–Strokorb / Forecaster's-dilemma point is the sharper framing, and it changes what I'd suggest: a denser tail grid is a wash — it trades numerical-integration error for the statistical estimation error you cited — so grid density isn't the lever. The real lever is the *estimand*: full CRPS vs a threshold-/quantile-weighted CRPS (Gneiting–Ranjan 2011). twCRPS stays strictly proper while emphasizing a declared region, but it answers a different question — and Brehmer–Strokorb is precisely why you can't fold "tail sensitivity" into a single CRPS number: there is no proper score sensitive to the tail alone. So the honest design is to *report both* — full CRPS plus a twCRPS at a declared threshold — rather than trying to make one CRPS more tail-aware through discretization. Keeps propriety, and makes the tail claim explicit instead of smuggled into the grid.

Happy to keep exchanging on support-point placement — I'll follow ScoringBench, and can sketch a twCRPS-at-threshold column against `metrics.py` if that's useful. Appreciate the offer to go deeper; glad to.

**Comment by jonaslandsgesell** [comment]
PS: We report left, right and center weighted CRPS for that reason. Still we feel making probabilisitic forecasts is valuable and a great enhancement for TFMs.

**Comment by Whatsonyourmind** [comment]
Left/right/center weighted — you're already there, then; that's the clean version of it. And agreed: getting TFMs to emit calibrated predictive distributions at all is the bigger unlock — the tail-weighting is just how you keep the eval honest on top of it. I'll keep following ScoringBench.

**Comment by Whatsonyourmind** [comment]
Following up on the twCRPS-at-threshold sketch I mentioned — and the key point up front, because it's what makes it *not* redundant with the `wcrps_left/right/center` you already ship.

Your `compute_quantile_wcrps` weights in **quantile space** (Gneiting & Ranjan 2011, Eq. 17: `2∫ρ_α v(α)dα`, with `v=(1−α)²/α²/α(1−α)`). Threshold-weighted CRPS is the *other* family in that same paper — it weights in **outcome space**:

```
twCRPS(F, y) = ∫ (F(z) − 1{y ≤ z})² w(z) dz,   canonical upper tail: w(z) = 1{z ≥ t}
```

The distinction is the whole value: a quantile weight emphasizes a *distribution-relative* level — the α=0.9 quantile is a different point on the real line for every forecast — whereas a threshold weight pins the emphasis to a **fixed, decision-relevant outcome level `t`** shared across all forecasts. That fixed-threshold framing is exactly the Brehmer–Strokorb (2019) / forecaster's-dilemma setting: "how well is the declared high-impact region predicted," comparable across heterogeneous forecasts. Strictly proper for any fixed `w ≥ 0`, and — as we agreed — reported **alongside** the full CRPS, never replacing it (there's no proper score sensitive to the tail alone).

It slots straight into your histogram/energy-β=1 idiom — same `(cdf − target_cdf)²`-integrated-over-`bin_widths` shape as `compute_crls`, so it's a Brier integral on the grid restricted to `z ≥ t`:

```python
target_cdf = (bin_idx >= y_bin[:, None]).to(cdf.dtype)      # step CDF of the point mass at y
brier_bins = (cdf - target_cdf) ** 2
t    = torch.quantile(y, q)                                 # fixed outcome threshold, e.g. q=0.90
w_up = (bin_mids >= t).to(cdf.dtype)                        # 1{z >= t}
twcrps_upper = (brier_bins * w_up * bin_widths).sum(-1).mean()
```

Two things I checked before proposing it: this histogram form matches the definitional chaining-form twCRPS (Allen–Ferro–Kwasniok 2023, `v(z)=(z−t)₊`) to ~1e-5, and at `w≡1` it reduces to your β=1 CRPS to bin-discretization (~7e-4, same convention as `compute_crls`). Defaulting `t` to upper quantiles of the pooled targets keeps the column comparable across datasets, and it auto-registers on the autorank leaderboard as lower-is-better with no other change.

Happy to open it as a scoped PR against `metrics.py` — one helper + the call in `_compute_scoring_rules_torch` + a couple of tests (reduction-to-CRPS and the tail ≤ full-CRPS bound) — or to take it to the private channel you offered, whichever you prefer. Your call on whether the fixed-threshold column earns its place next to the quantile-weighted ones.
